src/api/java/baritone/api/utils/command/BaritoneChatControl.java

```diff
@@ -0,0 +1,225 @@
+            msg = "help";
+        }
+
+        Pair<String, List<CommandArgument>> pair = CommandExecution.expand(msg);
+        String command = pair.first();
+        String rest = msg.substring(pair.first().length());
+        ArgConsumer argc = new ArgConsumer(pair.second());
+
+        if (!argc.has()) {
+            Settings.Setting setting = settings.byLowerName.get(command.toLowerCase(Locale.US));
+
+            if (setting != null) {
+                logRanCommand(command, rest);
+
+                if (setting.getValueClass() == Boolean.class) {
+                    CommandManager.execute(String.format("set toggle %s", setting.getName()));
+                } else {
+                    CommandManager.execute(String.format("set %s", setting.getName()));
+                }
+
+                return true;
+            }
+        } else if (argc.hasExactlyOne()) {
+            for (Settings.Setting setting : settings.allSettings) {
+                if (setting.getName().equals("logger")) {
+                    continue;
+                }
+
+                if (setting.getName().equalsIgnoreCase(pair.first())) {
+                    logRanCommand(command, rest);
+                    CommandManager.execute(String.format("set %s %s", setting.getName(), argc.getString()));
+                    return true;
+                }
+            }
+        }
+
+        CommandExecution execution = CommandExecution.from(pair);
+
+        if (isNull(execution)) {
+            return false;
+        }
+
+        logRanCommand(command, rest);
+        CommandManager.execute(execution);
+
+        return true;
+    }
+
+    @Override
+    public void onPreTabComplete(TabCompleteEvent.Pre event) {
+        if (!settings.prefixControl.value) {
+            return;
+        }
+
+        String prefix = event.prefix.get();
+        String commandPrefix = settings.prefix.value;
+
+        if (!prefix.startsWith(commandPrefix)) {
+            return;
+        }
+
+        String msg = prefix.substring(commandPrefix.length());
+
+        List<CommandArgument> args = CommandArgument.from(msg, true);
+        Stream<String> stream = tabComplete(msg);
+
+        if (args.size() == 1) {
+            stream = stream.map(x -> commandPrefix + x);
+        }
+
+        event.completions.set(stream.toArray(String[]::new));
+    }
+
+    public Stream<String> tabComplete(String msg) {
+        List<CommandArgument> args = CommandArgument.from(msg, true);
+        ArgConsumer argc = new ArgConsumer(args);
+
+        if (argc.hasAtMost(2)) {
+            if (argc.hasExactly(1)) {
+                return new TabCompleteHelper()
+                        .addCommands()
+                        .addSettings()
+                        .filterPrefix(argc.getString())
+                        .stream();
+            }
+
+            Settings.Setting setting = settings.byLowerName.get(argc.getString().toLowerCase(Locale.US));
+
+            if (nonNull(setting)) {
```

Comment:
why is this a thing. can you please just use `!= null` like a normal person this is sickening.